rfd/xxxx-bot-for-cert-renewals.md

```diff
@@ -0,0 +1,626 @@
+complicated use cases can make use of a `tbot.yaml` configuration file.
+
+An example `tbot.yaml` with all advanced options specified:
+
+```yaml
+# Storage for the bot's private data
+storage:
+  # can also support any backend supporting both store + fetch
+  directory: /var/lib/tbot
+
+sinks:
+  - directory: /home/alice/.tbot
+    roles: [a, b, c]
+
+  # TODO: probably need to specify permissions (this will be ugly)
+  - directory: /home/bob/.tbot
+    roles: [b, c]
+
+  # k8s secrets support store and fetch
+  - kubernetesSecret:
+      kubeconfig: /home/user/kubeconfig
+      namespace: example
+      secretName: example-secret
+    roles: [c]
+
+  # Webhooks support only store
+  - webhook: http://localhost:8001
+    roles: [a]
+```
+
+#### External Configuration Assist
+
+In order to make it easy to consume certificates generated by `tbot`, the agent
+will have a `tbot config` subcommand that can render configuration snippets
+for common integration points.
+
+For example:
+
+```
+$ tbot start ...
+$ tbot config --ssh
+To configure sshd to trust this host certificate, add the following to your sshd_config:
+
+HostCertificate /foo/bar/tbot-cert.pub
+```
+
+(In the above, the instructional text will be written to stderr, so that the following
+will be possible: `tbot config --ssh | sudo tee -a /etc/ssh/sshd_config`)
+
+If `tbot` is managing multiple `--cert`s, you can pass `mode:destination`
+to tell it which certificate to render configuration for.
+
+Note: `tctl auth sign` has a tiny amount of configuration templating already for
+Postgres and MongoDB certificates. We should aim to move this logic out into a
+separate package that can render configuration snippets for a variety of
+external systems. This work is out of scope for the purposes of this RFD.
+
+#### Polling for expiration
+
+In order to watch a certificate for expiration, the
+`teleport/lib/utils/interval` package will be used. Note that tbot will only set
+up polling for _user certs_, as host certificates issued by teleport do not
+expire (host certificates *will* be re-issued during CA rotations).
+
+#### Initial User Certificates
+
+In order for the bot to obtain its first set of user certificates, a new
+endpoint will be added to the API server at `tokens/register/user`. This
+endpoint will mimic the host registration flow, accepting a pre-generated token
+and a public key to sign, but will use user tokens instead of provisioning
+tokens.
+
+Today, user tokens can be used to initiate the user invite flow, handle password
+resets, and manage the account recovery process. In order to generate initial
+user certificates, we will introduce a new type of user token for certificate
+renewal bots.
+
+When the backend receives the request to generate new user credentials, it validates
+the token, ensuring that the token is of the correct type and has not expired.
+
+*TODO (Tim): Investigate adding a new gRPC endpoint for this functionality instead
```

Comment:
I agree, this should be a separate GRPC endpoint, probably integrated with AWS node joining? Can we modify the existing GRPC endpoint to support multiple use-cases?

Reply:
I've now migrated this to a pure gRPC endpoint on the tbot branch (see the [`timothyb89/tbot`](https://github.com/gravitational/teleport/tree/timothyb89/tbot) branch).

As for AWS node joining, I see no reason why the endpoint (specifically the `RenewableCertsRequest` message) couldn't be extended to accept an EC2 identity document as we do with existing node joining.